spec/fixtures/generate_external_certs.sh

```diff
@@ -0,0 +1,337 @@
+  openssl req -sha256 -new -key $DIR_KEYS/ca_$1.key -out $CSR_DIR/ca_$1.csr -subj "$3"
+  openssl ca -batch \
+    -config $DIR_CONF/ca_$2.conf \
+    -extensions v3_ca \
+    -out $DIR_CERTS/ca_$1.crt \
+    -infiles $CSR_DIR/ca_$1.csr 2>/dev/null
+}
+
+function append_altnames() { # $1 altnames, $2 filename
+  IFS='-' read -r -a altnames <<< "$1"
+  for index in "${!altnames[@]}"
+  do
+      echo "DNS.$(($index+1)) = ${altnames[index]}" >> $2
+  done
+}
+
+function append_ip_altnames() { # $1 altnames, $2 filename
+  IFS='-' read -r -a altnames <<< "$1"
+  for index in "${!altnames[@]}"
+  do
+      echo "IP.$(($index+1)) = ${altnames[index]}" >> $2
+  done
+}
+
+function site_request_conf() { # $1 file, $2 bits
+  echo "[req]" > $1
+  echo default_bits = $2 >> $1
+  echo distinguished_name = req_distinguished_name >> $1
+  echo req_extensions = req_ext >> $1
+  
+  # Various prompts for the portions of the subject would go here, as well as defaults.
+  # They can, though, be left blank when the subject is provided on the command line.
+  cat >>$1 <<EOL
+[req_distinguished_name]
+EOL
+  
+  echo -e "[req_ext]\nsubjectAltName = @alt_names\n" >> $1
+  echo -e "[alt_names]" >> $1
+}
+
+function gen_start_end() { # $1 good/expired
+  # Right now we are just using harded good/bad for the date range validity.
+  # For most purposes it would be better to expose the clean text specification
+  # of both start and end as parameters.
+  if [ $1 = "good" ]
+  then
+    START_DATE=$(TZ=UTC date +"%Y%m%d%H%M%SZ" -d "-1 day")
+    END_DATE=$(TZ=UTC date +"%Y%m%d%H%M%SZ" -d "+5 years")
+  else
+    START_DATE=$(TZ=UTC date +"%Y%m%d%H%M%SZ" -d "-10 day")
+    END_DATE=$(TZ=UTC date +"%Y%m%d%H%M%SZ" -d "-5 days")
+  fi
+}
+
+function create_site_crt() { # $1 domain, $2 ca, $3 message_digest, $4 rsa_bits, $5 good/expired, $6 subject, $7 altnames, $8 ipaltnames
+  echo "Generating site key \"$1\""
+  openssl genrsa -out $DIR_KEYS/site_$1.key $4 2>/dev/null
+  
+  echo "Generating site CRT \"$1\""
+  
+  FILE=/tmp/ca_gen/site_$1.req.conf
+  site_request_conf $FILE $4
+  if [ "$7" != "0" ]; then append_altnames    $7 $FILE; fi
+  if [ "$8" != "0" ]; then append_ip_altnames $8 $FILE; fi
+  
+  openssl req -new -sha256 -key $DIR_KEYS/site_$1.key -subj "$6" -out $CSR_DIR/site_$1.csr -config $FILE
+  # It is possible to specify altnames the following way instead of constructing a conf file.
+  # There are additional ways beyond this using new features of latest openssl as well.
+  #openssl req -new -sha256 \
+  #  -key keys/site_$1.key \
+  #  -subj "$6" \
+  #  -reqexts SAN \
+  #  -config <(cat /etc/ssl/openssl.cnf \
+  #   <(printf "\n[SAN]\nsubjectAltName=DNS:example.com,DNS:www.example.com")) \
+  #  -out $CSR_DIR/site_$1.csr
+  
+  # Cleanest way to see the actual contents of the CSR to verify it was created correctly.
+  #openssl req -in $CSR_DIR/site_$1.csr -text -noout
+  
+  # Unclean very raw way to see the contents of the CSR
```

Comment:
This is barely if ever useful; the openssl text output is much better so this mentioned method could be deleted.

Reply:
Removed the unclean raw way comment and line of code.